sql/parser/type_check.go

```diff
@@ -0,0 +1,388 @@
+
+	case *ExistsExpr:
+		// The subquery within the exists should have been executed before
+		// expression evaluation and the exists nodes replaced with the result.
+
+	case BytesVal:
+		return stringType, nil
+
+	case StrVal:
+		return stringType, nil
+
+	case IntVal:
+		return intType, nil
+
+	case NumVal:
+		return floatType, nil
+
+	case BoolVal:
+		return boolType, nil
+
+	case ValArg:
+		// Placeholders should have been replaced before expression evaluation.
+
+	case *QualifiedName:
+		return nil, fmt.Errorf("qualified name \"%s\" not found", t)
+
+	case Tuple:
+		// TODO(tamird): handling of tuples leaves much to be desired. This
+		// is largely a consequence of go's type system not being
+		// expressive enough to encode the type of a heterogenous tuple
+		// (dBool, dString) as distinct from the tuple (dBool). We'll have
+		// to write our own type system here to do this properly.
+		return tupleType, nil
+
+	case DReference:
+		return reflect.TypeOf(t.Datum()), nil
+
+	case Datum:
+		return reflect.TypeOf(t), nil
+
+	case *Subquery:
+		// The subquery should have been executed before expression evaluation and
+		// the result placed into the expression tree.
+
+	case *BinaryExpr:
+		return typeCheckBinaryExpr(t)
+
+	case *UnaryExpr:
+		return typeCheckUnaryExpr(t)
+
+	case *FuncExpr:
+		return typeCheckFuncExpr(t)
+
+	case *CaseExpr:
+		return typeCheckCaseExpr(t)
+
+	case *CastExpr:
+		return typeCheckCastExpr(t)
+	}
+
+	return nil, util.Errorf("eval: unexpected expression: %T", expr)
+}
+
+func typeCheckBooleanExprs(exprs ...Expr) (reflect.Type, error) {
+	for _, expr := range exprs {
+		exprType, err := TypeCheckExpr(expr)
+		if err != nil {
+			return nil, err
+		}
+		if exprType != boolType {
+			return nil, fmt.Errorf("incompatible AND argument type %s", typeToValue[exprType].Type())
+		}
+	}
+	return boolType, nil
+}
+
+// TODO(tamird): reduce duplication with evalComparisonExpr.
+func typeCheckComparisonExpr(expr *ComparisonExpr) (reflect.Type, error) {
+	leftType, err := TypeCheckExpr(expr.Left)
+	if err != nil {
+		return nil, err
+	}
+	rightType, err := TypeCheckExpr(expr.Right)
+	if err != nil {
+		return nil, err
+	}
+
+	return typeCheckComparisonOp(expr.Operator, leftType, rightType)
+}
+
+func typeCheckComparisonOp(op ComparisonOp, leftType, rightType reflect.Type) (reflect.Type, error) {
+	switch op {
+	case NE:
+		// NE(left, right) is implemented as !EQ(left, right).
+		op = EQ
+	case GT:
+		// GT(left, right) is implemented as LT(right, left)
+		op = LT
+		leftType, rightType = rightType, leftType
+	case GE:
+		// GE(left, right) is implemented as LE(right, left)
+		op = LE
+		leftType, rightType = rightType, leftType
+	case NotIn:
+		// NotIn(left, right) is implemented as !IN(left, right)
+		op = In
+	}
+
+	// TODO(pmattis): Memoize the cmpOps lookup as we've done for unaryOps and
+	// binOps.
+	if f, ok := cmpOps[cmpArgs{op, leftType, rightType}]; ok {
+		return reflect.TypeOf(f).Out(0), nil
```

Comment:
yeah, but I felt this was nicer than duplicating that fact (which is why i changed the type of cmpOps to return DBool rather than Datum)